File: Snake/settings/FileManager.java
package settings;

import java.awt.Font;
import java.awt.GridLayout;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.PrintWriter;

import javax.swing.JLabel;
import javax.swing.JPanel;

import elements.Block;
import elements.Cell;

public class FileManager {
	private String highScorePath;
	private String labyrinthPath;
	
	private Player[] scores;
	
	public static final String HIGHSCORE_PATH = "files\\high scores\\";
	public static final String HIGHSCORE_EXT = ".high";
	public static final String LABYRINTH_PATH = "files\\labyrinths\\";
	public static final String LABYRINTH_EXT = ".labyrinth";
	public static final String NOLABYRINTH = "noLabyrinth";
	
		
	public FileManager(String labyrinthName){		
		if(labyrinthName == null){
			highScorePath = HIGHSCORE_PATH + NOLABYRINTH + HIGHSCORE_EXT;
			labyrinthPath = null;
		}
		else{
			highScorePath = HIGHSCORE_PATH + labyrinthName + HIGHSCORE_EXT;
			labyrinthPath = LABYRINTH_PATH + labyrinthName + LABYRINTH_EXT;
		}
		
		scores = new Player[5];
		
		loadHighScores();
	}
	
	public boolean isNewHighScore(int points){
		if(points <= 0){return false;}		
		
		for( int i = 0; i < scores.length; i++){
			if(scores[i] == null){return true;}
			if(scores[i].getScore() < points){return true;}
		}
		
		return false;
	}
	
	public void addNewScore(String name, int points){		
		int lastIdx = scores.length - 1;
		for( int i = 0; i < scores.length; i++){
			if(scores[i] == null){lastIdx = i; break;}
		}
		
		for( int i = 0; i <= lastIdx; i++){
			if(scores[i] == null){scores[i] = new Player(name, points); break;}
			
			if(scores[i].getScore() < points){				
				for(int i2 = lastIdx; i2 > i; i2--){
					scores[i2] = scores[i2 - 1];  
				}
				scores[i] = new Player(name, points);
				break;
			}
		}
		
		saveHighScores();
	}
	
	private void saveHighScores(){
		PrintWriter pw;		
		
		try{			
		pw = new PrintWriter(highScorePath);
				
		for(int i = 0; i < scores.length; i++){
			if(scores[i] == null){break;}
			pw.println(scores[i]);
		}
		
		pw.close();
		
		}catch(IOException e){System.exit(0);}		
	}	
	
	private void loadHighScores(){
		BufferedReader br;
		
		try{
		br= new BufferedReader(new FileReader(highScorePath));		
				
		String s;
		int idx = 0;
		scores = new Player[5];
		
		while((s = br.readLine()) != null){
			scores[idx++] = new Player(
					(s.split(Player.SEPARATOR))[0],
					Integer.parseInt((s.split(" - "))[1]) );
		}
		
		br.close();		
		}catch(FileNotFoundException e){
			File f = new File(highScorePath);
			try{f.createNewFile();}catch(IOException ioe){}
		}
		catch(IOException e){System.exit(0);}
	}
	
	public void loadLabyrinth(Cell[][] c){
		if(labyrinthPath != null){loadLabyrinth(c, labyrinthPath);}		
	}
	
	public static void loadLabyrinth(Cell[][] c, String labyrinthPath){
		BufferedReader br;
		
		try{		
		br = new BufferedReader(new FileReader(labyrinthPath));

		String s;
		String[] xy;
		
		Block b = new Block();

		while((s = br.readLine()) != null){
			xy = s.split(",");
			
			int x = Integer.parseInt( xy[0] );
			int y = Integer.parseInt( xy[1] );
			
			c[x][y].setFigure(b);			
		}
		
		br.close();
		}catch(IOException e){System.exit(0);;}
	}
	
	public JPanel getDisplayHighScore(){
		JPanel d = new JPanel();		
		d.setLayout(new GridLayout(6, 1));
		
		String s = "";
		
		if(labyrinthPath != null){
			char[] aux = labyrinthPath.toCharArray();
		
			for(int i = labyrinthPath.lastIndexOf(".") - 1; i >= 0; i--){
				if( aux[i] == '\\' ) break;
				s = aux[i] + s;
			}
		}else{
			s = "No Labyrinth";
		}
		
		JLabel l = new JLabel( s, JLabel.CENTER);
		l.setFont(new Font("Comic Sans MS", Font.BOLD, 14));		
		d.add(l);
		
		for( int i = 0; i < scores.length; i++){
			if(scores[i] == null){
				d.add(new JLabel((i + 1) + "º -  ------------------", JLabel.CENTER));
			}else{
				d.add(new JLabel((i + 1) + "º - " +scores[i].toString(), JLabel.CENTER));
			}
		}
		
		return d;
	}
}

File: src/main/kotlin/org/snake/settings/FileManager.kt
package org.snake.settings

import org.snake.models.Block
import org.snake.models.GameMatrix
import org.snake.models.basic.Point
import org.snake.settings.Player
import org.snake.utils.*
import org.snake.views.BlockView
import java.awt.Font
import java.awt.GridLayout
import java.io.*
import javax.swing.JLabel
import javax.swing.JPanel
import kotlin.system.exitProcess

class FileManager(labyrinthName: String?) {
    private var highScorePath: String
    private var labyrinthPath: String? = null
    private var scores: FList<Player> = FNil

    init {
        if (labyrinthName == null) {
            highScorePath = HIGHSCORE_PATH + NOLABYRINTH + HIGHSCORE_EXT
            labyrinthPath = null
        } else {
            highScorePath = HIGHSCORE_PATH + labyrinthName + HIGHSCORE_EXT
            labyrinthPath = LABYRINTH_PATH + labyrinthName + LABYRINTH_EXT
        }
        loadHighScores()
    }

    fun isNewHighScore(points: Int): Boolean {
        if (points <= 0) {
            return false
        }
        return !scores.take(5).forall { it.score > points }
    }

    fun addNewScore(name: String, points: Int) {
        scores.insertBefore(Player(name, points)) { it.score < points }
        saveHighScores()
    }

    private fun saveHighScores() {
        PrintWriter(highScorePath).use { pw ->
            scores.foreach { pw.println(it) }
        }
    }

    private fun loadHighScores() {
        try {
            BufferedReader(FileReader(highScorePath)).use { br ->
                br.readLines().forEach { line ->
                    scores += Player(
                            line.split(Player.Companion.SEPARATOR).toTypedArray().get(0),
                            line.split(" - ").toTypedArray()[1].toInt()
                    )
                }
            }
        } catch (e: FileNotFoundException) {
            val f = File(highScorePath)
            try {
                f.createNewFile()
            } catch (ioe: IOException) {
            }
        } catch (e: IOException) {
            exitProcess(0)
        }
    }

    fun loadLabyrinth(c: GameMatrix) {
        if (labyrinthPath != null) {
            loadLabyrinth(c, labyrinthPath!!)
        }
    }

    val displayHighScore: JPanel
        get() {
            val d = JPanel()
            d.layout = GridLayout(6, 1)
            var s = ""
            val path = labyrinthPath
            if (path != null) {
                val aux = path.toCharArray()
                for (i in path.lastIndexOf(".") - 1 downTo 0) {
                    if (aux[i] == '\\') break
                    s = aux[i].toString() + s
                }
            } else {
                s = "No Labyrinth"
            }
            val l = JLabel(s, JLabel.CENTER)
            l.font = Font("Comic Sans MS", Font.BOLD, 14)
            d.add(l)

            scores.mapIndexed { index: Int, player: Player ->
                JLabel("${index + 1}* - ${player.score}", JLabel.CENTER)
            }.paddingIndexed(5) { index: Int ->
                JLabel("${index + 1}* -  ------------------", JLabel.CENTER)
            }.foreach { d.add(it) }

            return d
        }

    companion object {
        const val HIGHSCORE_PATH = "high-scores/"
        const val HIGHSCORE_EXT = ".high"
        const val LABYRINTH_PATH = "labyrinths/"
        const val LABYRINTH_EXT = ".labyrinth"
        const val NOLABYRINTH = "noLabyrinth"

        fun loadLabyrinth(c: GameMatrix, labyrinthPath: String) {
            BufferedReader(FileReader(labyrinthPath)).use { br ->
                var stringBuilder: StringBuilder = StringBuilder()
                br.readLines().forEach { line ->
                    val xy = line.split(",").toTypedArray()
                    val point = Point(xy[0].toInt(), xy[1].toInt())
                    val b = BlockView(Block(point))
                    c.setView(b, point.x, point.y)
                }
            }
        }
    }
}